presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java

```diff
@@ -666,6 +671,17 @@ public void setRoutines(List<RoutineInfo> routines)
         this.routines.set(ImmutableList.copyOf(routines));
     }
 
+    private synchronized DynamicFiltersStats getDynamicFiltersStats()
```

Comment:
I avoided `AtomicReference<Supplier<DynamicFiltersStats>>` because I think there is a potential for race condition with that. Example 
In `getQueryStats`, we get reference to the non-final df stats supplier, but the `get` on supplier isn't executed yet.
Query finishes, `unregisterDynamicFilteringQuery` is called, final df stats supplier is set and `dynamicFilterService.removeQuery` runs.
Now we continue with execution of `getQueryStats`, we still have reference to old supplier and calling get on it will give empty stats now as the dynamic filter service has cleaned up its maps.

Reply:
that makes sense, thanks